network/netcore/src/multiplexing/mod.rs

```diff
@@ -14,48 +14,52 @@
 //! [`TcpStream`]: tokio::net::tcp::TcpStream
 //! [`Yamux`]: crate::multiplexing::yamux::Yamux
 
+use async_trait::async_trait;
 use futures::{
-    future::Future,
     io::{AsyncRead, AsyncWrite},
     stream::Stream,
 };
 use std::{fmt::Debug, io};
 
 pub mod yamux;
 
+#[async_trait]
+pub trait Control {
+    /// The type of substreams opened by this Multiplexer.
+    ///
+    /// Must implement both AsyncRead and AsyncWrite.
+    type Substream: AsyncRead + AsyncWrite + Send + Debug + Unpin;
+
+    /// Requests that a new Substream be opened on the underlying connection.
+    async fn open_stream(&mut self) -> io::Result<Self::Substream>;
+
+    /// Close and shutdown this connection.
+    ///
+    /// After the returned future has resolved this multiplexer will be shutdown.  All subsequent
+    /// reads or writes to any still existing handles to substreams opened through this multiplexer
+    /// must return EOF (in the case of a read), or an error.
+    async fn close(&mut self) -> io::Result<()>;
+}
+
 /// A StreamMultiplexer is responsible for multiplexing multiple [`AsyncRead`]/[`AsyncWrite`]
 /// streams over a single underlying [`AsyncRead`]/[`AsyncWrite`] stream.
 ///
 /// New substreams are opened either by [listening](StreamMultiplexer::listen_for_inbound) for
 /// inbound substreams opened by the remote side or by [opening](StreamMultiplexer::open_outbound)
 /// and outbound substream locally.
+#[async_trait]
 pub trait StreamMultiplexer: Debug + Send + Sync {
```

Comment:
Send + Sync are intuitively needed because we would want to pass Muxers from one thread to another. Debug is needed because of a weird trait bound in `common/channel/src/lib.rs:130` where we log an element if it times out while in the queue. We should probably get rid of this, but in another PR.